src/enzo/ActiveParticleFindAll.C

```diff
@@ -0,0 +1,258 @@
+    buffer_size, LocalNumberOfActiveParticlesOnThisLevel, element_size, count, 
+    offset;
+  ActiveParticleList<ActiveParticleType> LocalActiveParticlesOfThisType, 
+    LocalActiveParticlesOnThisLevel, ParticlesOnThisProc;
+
+  HierarchyEntry **Grids = NULL;
+  int NumberOfGrids, *NumberOfActiveParticlesInGrids = NULL;
+  ActiveParticleType_info *ap_info = NULL;
+
+  *GlobalNumberOfActiveParticles = 0;
+  LocalNumberOfActiveParticles = 0;
+
+  for (type = 0; type < EnabledActiveParticlesCount; type++) {
+      
+    ap_info = EnabledActiveParticles[type];
+    ap_id = ap_info->GetEnabledParticleID();
+    
+    if (ap_id == ActiveParticleIDToFind) {
+      
+      /* Traverse the hierarchy and generate a buffer of all of the active 
+	 particles of this type on this processor */
+      
+      for (level = 0; level <= MaximumRefinementLevel; level++) {
+	NumberOfGrids = GenerateGridArray(LevelArray, level, &Grids);
+	NumberOfActiveParticlesInGrids = new int[NumberOfGrids];
+	LocalNumberOfActiveParticlesOnThisLevel = 0;
+
+	/* In a first pass, find the number of active particles on each grid */
+	for(GridNum = 0; GridNum < NumberOfGrids; GridNum++) {
+	  
+	  NumberOfActiveParticlesInGrids[GridNum] = Grids[GridNum]->GridData->
+	    ReturnNumberOfActiveParticlesOfThisType(ActiveParticleIDToFind);
+	  LocalNumberOfActiveParticlesOnThisLevel += NumberOfActiveParticlesInGrids[GridNum];
+	} /* ENDFOR grids */
+	
+	offset = 0;
+	
+	if (LocalNumberOfActiveParticlesOnThisLevel > 0) {
+
+	  /* If we've already found active particles, save the list */
+
+	  if (LocalNumberOfActiveParticles != 0) 
+	    {
+	      ParticlesOnThisProc = LocalActiveParticlesOfThisType;
+	      LocalActiveParticlesOfThisType.clear();
+	    }
+      
+      /* In a second pass, fill up the active particle list for this level*/
+	  for(GridNum = 0; GridNum < NumberOfGrids; GridNum++) {
+        // I've removed this function from the grid class, do this another way.
+	    Grids[GridNum]->GridData->
+	      AppendActiveParticlesToList(LocalActiveParticlesOnThisLevel,
+              ActiveParticleIDToFind);
+	    offset += NumberOfActiveParticlesInGrids[GridNum];
+	  } 
+
+	  LocalActiveParticlesOfThisType.reserve(
+          offset + LocalNumberOfActiveParticles);
+	  
+	  /* If we've already found active particles, copy the cached
+	     list to the new one and delete the old list */
+	  if (LocalNumberOfActiveParticles != 0) {
+	    for (i = 0; i < LocalNumberOfActiveParticles; i++)
+	      LocalActiveParticlesOfThisType.copy_and_insert(
+              *ParticlesOnThisProc[i]);
+	    ParticlesOnThisProc.clear();
+	  }
+
+	  /* Finally, append the new active particles to the list */
+	  for(i = LocalNumberOfActiveParticles; i < offset + LocalNumberOfActiveParticles; i++) 
+	    LocalActiveParticlesOfThisType.copy_and_insert(
+            *LocalActiveParticlesOnThisLevel[i - LocalNumberOfActiveParticles]);
```

Comment:
How is the Active Particle type accounted here?

Reply:
Oh - each ActiveParticleType calls this function separately. So when it returns it returns the number of that type. Is that what you meant?